src/decorators/safe-url.decorator.ts

```diff
@@ -0,0 +1,152 @@
+import {
+  registerDecorator,
+  ValidationOptions,
+  ValidationArguments,
+} from 'class-validator';
+import { Transform } from 'class-transformer';
+import * as sanitizeHtml from 'sanitize-html';
+
+export function IsSafeUrl(validationOptions?: ValidationOptions) {
+  return function (object: Object, propertyName: string) {
+    registerDecorator({
+      name: 'isSafeUrl',
+      target: object.constructor,
+      propertyName: propertyName,
+      options: validationOptions,
+      validator: {
+        validate(value: any, args: ValidationArguments) {
+          if (typeof value !== 'string') return false;
+
+          // Instant rejection of URLs containing the most dangerous characters
+          if (/[<>"'`]/.test(value)) {
+            return false;
+          }
+
+          try {
+            // Decoding URL before checking content
+            const decodedUrl = decodeURIComponent(value);
+
+            // Checking if the decoded URL contains dangerous characters
+            if (/[<>"'`]/.test(decodedUrl)) {
+              return false;
+            }
+
+            const url = new URL(value);
+
+            // Basic protocol check
+            if (!['http:', 'https:'].includes(url.protocol)) {
+              return false;
+            }
+
+            // Block Punycode
+            if (url.hostname.includes('xn--')) {
+              return false;
+            }
+
+            // Blocking credentials in URL
+            if (url.username || url.password || url.href.includes('@')) {
```

Comment:
**issue**: Blocking credentials in URLs is crucial, but consider logging these attempts for security audits.